### PR TITLE
Make sure default ALC is initialized before resolving satellite assembly

### PR DESCRIPTION
**Pull request by elinor-fung**
Fix dotnet/corefx#42477

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs [hunk 1]
```diff
@@ -589,6 +589,7 @@ public override Assembly GetSatelliteAssembly(CultureInfo culture, Version? vers
             // This stack crawl mark is never used because the requesting assembly is explicitly specified,
             // so the value could be anything.
             StackCrawlMark unused = default;
+            _ = AssemblyLoadContext.Default; // Ensure default ALC is initialized - the runtime may use it as part of satellite assembly resolution
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs]: Where does the runtime use it? What prevents it being initialized on demand - right when the runtime uses it?

Comment (elinor-fung) [reply src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs]: Mm, I could make `CLRPrivBinderCoreCLR` call up into some managed function that initializes the managed default ALC if it doesn't already have the managed ALC during an assembly bind. It tries to use the managed ALC for resolution if the assembly wasn't already loaded or in the TPA list: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/binder/clrprivbindercoreclr.cpp#L79-L82

Comment (jkotas) [reply src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs]: Yes, I think something like that would be better.

File: src/coreclr/src/binder/clrprivbindercoreclr.cpp [hunk 2]
```diff
@@ -71,12 +72,28 @@ HRESULT CLRPrivBinderCoreCLR::BindAssemblyByName(IAssemblyName     *pIAssemblyNa
...
+                SString &culture = pAssemblyName->GetCulture();
+                if (!culture.IsEmpty() && !culture.EqualsCaseInsensitive(g_BinderVariables->cultureNeutral))
+                {
+                    // Make sure the managed default ALC is initialized.
+                    GCX_COOP();
+                    PREPARE_NONVIRTUAL_CALLSITE(METHOD__ASSEMBLYLOADCONTEXT__INITIALIZE_DEFAULT_CONTEXT);
+                    DECLARE_ARGHOLDER_ARRAY(args, 0);
+                    CALL_MANAGED_METHOD_NORET(args)
+
+                    pManagedAssemblyLoadContext = GetManagedAssemblyLoadContext();
```
Comment (jkotas) [inline src/coreclr/src/binder/clrprivbindercoreclr.cpp]: Nit: `_ASSERTE(pManagedAssemblyLoadContext  != NULL);` ?

**Review by jkotas — APPROVED** [review]

**Pull request merged**